compiler/rustc_hir_typeck/src/mem_categorization.rs

```diff
@@ -582,7 +582,8 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         match ty.kind() {
             ty::Tuple(args) => Ok(args.len()),
             _ => {
-                self.tcx().dcx().span_bug(span, "tuple pattern not applied to a tuple");
+                self.tcx().dcx().span_delayed_bug(span, "tuple pattern not applied to a tuple");
+                Err(())
```

Comment:
general thought, could `McResult` get changed to `Err(ErrorGuaranteed)`?

Reply:
I'll take a look once I finish cleaning up the rest of this mess :)

Reply:
I just looked, it would require adding several `span_delayed_bug` calls in other places, which is doable. But the `McResult`s seem to all end up being returned from `cat_expr` and `cat_expr_unadjusted`, which are called from `ExprUseVisitor` via the `return_if_err!` macro which just does an early return in the `Err` case. So I don't think anything useful would be gained.

Reply:
assuming that mem categorization is necessary for something (which tbh, I am not 100% confident in), wouldn't it be good to make sure we only do an early return if there has been an actual error :thinking:

if we accidentally early return and fail to emit an error later on, that seems bad to me and exactly like the thing `ErrorGuaranteed` is supposed to prevent.